kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/SniAwareClusterEndpointConfigProvider.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal.clusterendpointprovider;
+
+import java.util.Locale;
+import java.util.regex.Pattern;
+
+import io.kroxylicious.proxy.config.BaseConfig;
+import io.kroxylicious.proxy.service.ClusterEndpointConfigProvider;
+import io.kroxylicious.proxy.service.HostPort;
+
+public class SniAwareClusterEndpointConfigProvider implements ClusterEndpointConfigProvider {
+
+    private static final EndpointMatchResult BOOTSTRAP_MATCHED = new EndpointMatchResult(true, null);
+    private static final EndpointMatchResult NO_MATCH = new EndpointMatchResult(false, null);
+    private static final Pattern NODE_ID_TOKEN = Pattern.compile("\\$\\(nodeId\\)");
+    public static final String LITERAL_NODE_ID = "$(nodeId)";
+    private final HostPort bootstrapAddress;
+    private final String brokerAddressPattern;
+    private final Pattern brokerAddressPatternRegExp;
+
+    public SniAwareClusterEndpointConfigProvider(SniAwareClusterEndpointProviderConfig config) {
+        this.bootstrapAddress = config.bootstrapAddress;
+        this.brokerAddressPattern = config.brokerAddressPattern;
+        this.brokerAddressPatternRegExp = config.brokerAddressPatternRegExp;
+    }
+
+    @Override
+    public HostPort getClusterBootstrapAddress() {
+        return this.bootstrapAddress;
+    }
+
+    @Override
+    public HostPort getBrokerAddress(int nodeId) {
+        if (nodeId < 0) {
+            // nodeIds of < 0 have special meaning to kafka.
+            throw new IllegalArgumentException("nodeId cannot be less than zero");
+        }
+        return new HostPort(brokerAddressPattern.replace(LITERAL_NODE_ID, Integer.toString(nodeId)), bootstrapAddress.port());
```

Comment:
I wonder if we should cache the replaced patterns keyed by node ID for efficiency.

Reply:
I added a comment to consider a LRU cache.  However, in the work that is planned for https://github.com/kroxylicious/kroxylicious/issues/271 it will be the EndpointRegistry that receives these queries, so the comment might ultimately belong elsewhere.